src/libponyc/pass/refer.c

```diff
@@ -1371,6 +1371,60 @@ static bool refer_seq(pass_opt_t* opt, ast_t* ast)
   return true;
 }
 
+// This function checks if the passed AST node represent
+// an explicit constructor call, and assume the
+// passed AST node is of TK_CALL type.
+//
+// When the passed receiver is a primitive, this returns
+// false.
+static bool is_creating_object_explicitly(ast_t *ast) {
+  pony_assert(ast_id(ast) == TK_CALL);
+
+  // The following diagram represent the AST structure
+  // of the following example:
+  //
+  //     C.create()
+  //
+  // type: TK_CALL (`C.create`)
+  // children:
+  // - type: TK_DOT
+  //   children:
+  //   - type: TK_TYPEREF (`C`)
+  //     name: C
+  //     children: [...]
+  //   - type: TK_ID (`create`)
+  //     name: create
+
+  ast_t* dot_expression = ast_child(ast);
+  if(ast_id(dot_expression) != TK_DOT)
+  {
+    return false;
+  }
+
+  ast_t* receiver = ast_child(dot_expression);
+  ast_t* message = ast_sibling(receiver);
+  pony_assert(ast_id(message) == TK_ID);
+
+  if(ast_id(receiver) != TK_TYPEREF)
+  {
+    return false;
+  }
+
+  ast_t* type = (ast_t*) ast_data(receiver);
+  if(ast_id(type) == TK_PRIMITIVE)
+  {
+    // Even if we can run `create` over primitives,
+    // this will not create new objects.
+    return false;
+  }
+  if(strcmp(ast_name(message), "create"))
```

Comment:
Not every constructor has the name "create", and it is possible to create a function whose name is "create". So the name is not the right criterion to check here.

Instead of checking here based on the name of the called method, we should be checking if it refers to a constructor on the type.

Here in the `refer.c` pass, that would be a bit complicated (though still possible).

My suggestion is that you move this entire function to the `verify.c` pass, as well as the entire `valid_is_comparand` function and the `refer_is` function that calls it (which would become `verify_is`).

The reason is that in the `expr.c` pass we actually resolve every `TK_DOT` to be specific to what it is calling. So each `TK_DOT` that calls a constructor gets turned into a `TK_NEWREF` or `TK_NEWBEREF` (for the case of an actor constructor).

So if you move this set of checks into the `verify.c` pass, it becomes simple to check if the call is to a constructor - just check the AST ID for `TK_NEWREF` or `TK_NEWBEREF`. And in that case you don't need to bother checking the left side (the type reference) so it becomes a bit simpler than you have here, which is nice.

Also, conceptually, this check doesn't really fit the `refer.c` pass anyway, as it has nothing to do with resolving `TK_REFERENCE` nodes. Instead, it belongs in the `verify` pass, which has a variety of post-type-analysis correctness checks.

Here is the link to the original commit where these checks were added, in case it helps you see which parts need to be moved to `verify.c`: https://github.com/ponylang/ponyc/commit/f37e281c49a85ee8df27033f2d707da3d52b1525

Reply:
Nice idea bringing this to the verify step. Let me try changing this PR with your suggestions. Thanks @jemc @SeanTAllen